src/ol/source/mapserver.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * @module ol/source/mapserver
+ */
+
+import {decode} from '../Image.js';
+import {getHeight, getWidth} from '../extent.js';
+import {appendParams} from '../uri.js';
+import {getRequestExtent} from './Image.js';
+
+/**
+ * @typedef {Object} LoaderOptions
+ * @property {string} url The MapServer url.
+ * @property {null|string} [crossOrigin] The `crossOrigin` attribute for loaded images.  Note that
+ * you must provide a `crossOrigin` value if you want to access pixel data with the Canvas renderer.
+ * See https://developer.mozilla.org/en-US/docs/Web/HTML/CORS_enabled_image for more detail.
+ * the image from the remote server.
+ * @property {number} [ratio=1] Ratio. `1` means image requests are the size of the map viewport, `2` means
+ * twice the width and height of the map viewport, and so on. Must be `1` or higher.
+ * @property {Object} [params] Additional query parameters.
+ * @property {function(HTMLImageElement, string): Promise<import('../DataTile.js').ImageLike>} [load] Function
+ * to perform loading of the image. Receives the created `HTMLImageElement` and the desired `src` as argument and
+ * returns a promise resolving to the loaded or decoded image. Default is {@link module:ol/Image.decode}.
+ */
+
+/**
+ * @param {string} baseUrl The MapServer url.
+ * @param {Object<string, string|number>} params Request parameters.
+ * @param {import("../extent.js").Extent} extent Extent.
+ * @param {import("../size.js").Size} size Size.
+ * @return {string} The MapServer map image request URL.
+ */
+function getUrl(baseUrl, params, extent, size) {
+  const width = Math.round(size[0]);
+  const height = Math.round(size[1]);
+  const mapSize = `${width} ${height}`;
+  const mapExt = `${extent[0]} ${extent[1]} ${extent[2]} ${extent[3]}`;
+
+  const baseParams = {
+    mode: 'map',
+    map_imagetype: 'png',
+    mapext: mapExt,
+    imgext: mapExt,
+    map_size: mapSize,
+    imgx: width / 2,
+    imgy: height / 2,
+    imgxy: mapSize,
+  };
+  Object.assign(baseParams, params);
+  return appendParams(baseUrl, baseParams);
+}
+
+/**
+ * Creates a loader for MapServer images.
```

Comment:
I think the documentation could be a little bit more verbose. At least as verbose as in the example, i.e. add a note that using WMS would be the preferred way to use a MapServer layer, and mention that this loader is for the CGI interface.